app/src/main/java/org/mozilla/focus/search/SearchEngineManager.java

```diff
@@ -5,27 +5,151 @@
+            final String languagePath = "search/" + Locales.getLanguage(locale);
+            final String defaultPath = "search/default";
+
+            final List<String> localeEngines = Arrays.asList(assetManager.list(localePath));
+            final List<String> languageEngines = Arrays.asList(assetManager.list(languagePath));
+            final List<String> defaultEngines = Arrays.asList(assetManager.list(defaultPath));
+
+            for (int i = 0; i < engineNames.length(); i++) {
+                final String engineName = engineNames.getString(i);
+                final String fileName = engineName + ".xml";
+
+                if (localeEngines.contains(fileName)) {
+                    searchEngines.add(SearchEngineParser.load(assetManager, localePath + "/" + fileName));
+                } else if (languageEngines.contains(fileName)) {
+                    searchEngines.add(SearchEngineParser.load(assetManager, languagePath + "/" + fileName));
+                } else if (defaultEngines.contains(fileName)) {
+                    searchEngines.add(SearchEngineParser.load(assetManager, defaultPath + "/" + fileName));
+                } else {
+                    Log.e(LOG_TAG, "Couldn't find configuration for engine: " + engineName);
+                }
+            }
+        } catch (IOException e) {
+            Log.e(LOG_TAG, "IOException while loading search engines", e);
+        } catch (JSONException e) {
+            throw new AssertionError("Reading search engine failed: ", e);
+        } finally {
+            this.searchEngines = searchEngines;
+
+            notifyAll();
+        }
+    }
+
+    private JSONArray loadSearchEngineListForLocale(Context context) throws IOException {
+        try {
+            final Locale locale = Locale.getDefault();
+            final JSONObject configuration = IOUtils.readAsset(context, "search/search_configuration.json");
+
+            // Try to find a configuration for the language tag first (de-DE)
+            final String languageTag = Locales.getLanguageTag(locale);
+            if (configuration.has(languageTag)) {
+                return configuration.getJSONArray(languageTag);
+            }
+
+            // Try to find a configuration for just the language (de)
+            final String language = Locales.getLanguage(locale);
+            if (configuration.has(language)) {
+                return configuration.getJSONArray(language);
+            }
+
+            // No configuration for the current locale found. Let's use the default configuration.
+            return configuration.getJSONArray("default");
+        } catch (JSONException e) {
+            // Assertion error because this shouldn't happen: We check whether a key exists before
+            // reading it. An error here would mean the JSON file is corrupt.
+            throw new AssertionError("Reading search configuration failed", e);
+        }
+    }
+
+    public synchronized List<SearchEngine> getSearchEngines() {
```

Comment:
We need to synchronize so that the main thread sees the data loaded by the background thread. In non synchronized code the main thread could see a cached version of `searchEngines` that is still null [[1]](https://www.securecoding.cert.org/confluence/display/java/Concurrency,+Visibility,+and+Memory).